js-sample/app.js

```diff
@@ -0,0 +1,304 @@
+
+    obj[key] = function(error) {
+      results[key] = {
+        error: error,
+        args: Array.prototype.slice.call(arguments, 1)
+      };
+
+      if (error) {
+        hasError = true;
+      }
+
+      originalCallback.apply(obj, arguments);
+      checkDone();
+    };
+  });
+};
+
+var bandwidthCalculatorObj = function(config) {
+  var intervalId;
+
+  config.pollingInterval = config.pollingInterval || 500;
+  config.windowSize = config.windowSize || 2000;
+  config.subscriber = config.subscriber || undefined;
+
+  return {
+    start: function(reportFunction) {
+      var statsBuffer = [];
+      var last = {
+        audio: {},
+        video: {}
+      };
+
+      intervalId = window.setInterval(function() {
+        config.subscriber.getStats(function(error, stats) {
+          var snapshot = {};
+          var nowMs = new Date().getTime();
+          var sampleWindowSize;
+
+          ['audio', 'video'].forEach(function(type) {
+            snapshot[type] = Object.keys(stats[type]).reduce(function(result, key) {
+              result[key] = stats[type][key] - (last[type][key] || 0);
+              last[type][key] = stats[type][key];
+              return result;
+            }, {});
+          });
+
+          // get a snapshot of now, and keep the last values for next round
+          snapshot.timestamp = stats.timestamp;
+
+          statsBuffer.push(snapshot);
+          statsBuffer = statsBuffer.filter(function(value) {
+            return nowMs - value.timestamp < config.windowSize;
+          });
+
+          sampleWindowSize = getSampleWindowSize(statsBuffer);
+
+          if (sampleWindowSize !== 0) {
+            reportFunction(calculatePerSecondStats(
+              statsBuffer,
+              sampleWindowSize
+            ));
+          }
+        });
+      }, config.pollingInterval);
+    },
+
+    stop: function() {
+      window.clearInterval(intervalId);
+    }
+  };
+};
+
+var performQualityTest = function(config, callback) {
+  var startMs = new Date().getTime();
+  var testTimeout;
+  var currentStats;
+
+  var bandwidthCalculator = bandwidthCalculatorObj({
+    subscriber: config.subscriber
+  });
+
+  var cleanupAndReport = function() {
+    currentStats.elapsedTimeMs = new Date().getTime() - startMs;
+    callback(undefined, currentStats);
+
+    window.clearTimeout(testTimeout);
+    bandwidthCalculator.stop();
+
+    callback = function() {};
+  };
+
+  // bail out of the test after 30 seconds
+  window.setTimeout(cleanupAndReport, config.timeout);
+
+  bandwidthCalculator.start(function(stats) {
+    console.log(stats);
+
+    // you could do something smart here like determine if the bandwidth is
+    // stable or acceptable and exit early
+    currentStats = stats;
+  });
+};
+
+var publisherEl = document.createElement('div');
+var subscriberEl = document.createElement('div');
+var session;
+var publisher;
+var subscriber;
+var statusContainerEl;
+var statusMessageEl;
+var statusIconEl;
+
```

Comment:
I recommend putting these variable declarations at the top of the file, along with the testStreamingCapability function and other code most closely related to the stream test. I'd move the helper functions (e.g., pluck(), sum(), etc.) to the end of the file.